Automattic/jetpack
# Licensing: Display notice when user has unactivated licenses

This PR displays a ""You have an inactive product. Activate now" notice in the dashboard if the user has unactivated product license keys available.

#### Changes proposed in this Pull Request:


- Register API route (/jetpack/v4/licensing/user/counts)  for fetching user license counts.
- Fetch user license counts and store them in initialState.
- Create reducer, actions, and selector for retrieving and updating user license counts in redux state.
- Display notice when `hasDetachedUserLicenses( state )` is `true`.
- Removes notice after a license key is activated (if `hasDetachedUserLicenses( state )` is `false`).

#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






- Spin up a site using this Jetpack branch. (If using a JN site, you'll want to select 'More options' and include the "Code Snippets" plugin because we'll need to add a filter)
- Connect Jetpack to your WordPress.com user account.
- Enable the license-key input box UI on the "My plan" page by adding `add_filter( 'jetpack_licensing_ui_enabled', '__return_true' );`
- Go to `/wp-admin/?page=jetpack#/dashboard#/my-plan` (aka "My plan" page)
- Verify you **do not see a notice** at the top that says, "You have an inactive product. Activate now." 
- Verify you can see the License Key input field (at the bottom of the "My Plan" section).

Now obtain a 'user' product license key:

- log into your wpcom sandbox, and in file `wp-content/mu-plugins/0-sandbox.php` add: `add_filter( 'jetpack_license_checkout_m2_enabled', '__return_true' );`. (Enabling this flag makes Jetpack checkout issue you a license when completing checkout).
- Point `public-api.wordpress.com` to your sandbox IP (in your hosts file).
- Open a new command-line window in your sandbox and `tail -f /tmp/php-errors`
- Make sure you are proxied and go to https://wordpress.com/checkout/jetpack/jetpack_scan (siteless checkout flow).
- Complete checkout (using credits is fine).
- View your error log, you should see the output: `Jetpack user license was created for subscription_id:17829922, belonging to user_id:78351100 and blog_id:198416894`, but with different id values corresponding to your user and site. And you'll also see, `License key is: jetpack_scan_monthly_oRhIEF4ckyVP6mCuHaD1vYnw5`, but with your unique key.
- Copy the license key from the error log output.
- un-sandbox `public-api.wordpress.com`.

Test that the notice appears now that you have been provisioned a detached license, and that it disappears when the license is attached (if you no longer have any detached licenses left):

- Go to your Jetpack site and reload `/wp-admin/?page=jetpack#/dashboard#/my-plan` (aka "My plan" page).
- Verify that you see a notice at the top that says, **"You have an inactive product. Activate now."**
- Paste the license key into the License Key input box and submit.
- Verify the "You have an inactive product. Activate now." notice **disappears** (Also you should see that the Jetpack Scan product has been provisioned for your site(You may need to refresh the page to see that))
